### DISCUSSION TITLE
How to connect Node-RED as express middleware? As custom extension.

### DISCUSSION THREAD

**Discussion by ivanq3w** [directus/directus#6992] [Feature Requests]
Here is a guide: https://nodered.org/docs/user-guide/runtime/embedding

We need http-server to be exposed:

```js
var http = require('http');
var express = require("express");
var RED = require("node-red");
// Create an Express app
var app = express();
// Add a simple route for static content served from 'public'
app.use("/",express.static("public"));
// Create a server
var server = http.createServer(app);
// Create the settings object - see default settings.js file for other options
var settings = {
    httpAdminRoot:"/red",
    httpNodeRoot: "/api",
    userDir:"/home/nol/.nodered/",
    functionGlobalContext: { }    // enables global context
};
// Initialise the runtime with a server and settings
RED.init(server,settings);
// Serve the editor UI from /red
app.use(settings.httpAdminRoot,RED.httpAdmin);
// Serve the http nodes UI from /api
app.use(settings.httpNodeRoot,RED.httpNode);
server.listen(8000);
// Start the runtime
RED.start();
```

The same situation is with socket.io:

```js
const express = require('express');
const app = express();
const http = require('http');
const server = http.createServer(app);

app.get('/', (req, res) => {
  res.send('<h1>Hello world</h1>');
});

server.listen(3000, () => {
  console.log('listening on *:3000');
});
```

**Comment by ivanq3w** [comment]
How to implement it with custom endpoints: https://docs.directus.io/guides/api-endpoints/

**Comment by rijkvanzanten** [comment]
There's a `server.start.before` async hook you can use to get access to the http server instance 👍🏻 

https://github.com/directus/directus/blob/afdeb980cb035192b1075d8567080aea921198da/api/src/start.ts#L17

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
